core/iwasm/aot/aot_runtime.c

```diff
@@ -783,12 +783,18 @@ static AOTMemoryInstance *
 memory_instantiate(AOTModuleInstance *module_inst, AOTModuleInstance *parent,
                    AOTModule *module, AOTMemoryInstance *memory_inst,
                    AOTMemory *memory, uint32 memory_idx, uint32 heap_size,
-                   char *error_buf, uint32 error_buf_size)
+                   uint32 max_memory_pages, char *error_buf,
+                   uint32 error_buf_size)
 {
     void *heap_handle;
     uint32 num_bytes_per_page = memory->num_bytes_per_page;
     uint32 init_page_count = memory->mem_init_page_count;
-    uint32 max_page_count = memory->mem_max_page_count;
+    uint32 max_page_count =
+        max_memory_pages == 0
+            ? memory->mem_max_page_count
```

Comment:
How about handling `max_memory_pages == 0` inside `wasm_runtime_get_overwritten_max_mem`, and here it just calls `wasm_runtime_get_overwritten_max_mem`?